### PR TITLE
Service unavailable page shown when api returns a 404

### PR DESCRIPTION
**Pull request by MartinBentley**
### Context
This is the bug where the user is presented with a "Service Down" page when trying to access a non-existent course. Here is the trello card:
https://trello.com/c/cjI2ies3/575-find-showing-503-service-unavailable-when-the-api-returns-404

In the http client, the HttpClientWrapper class in the nuget package was generating an exception on all api calls where the status is not 200. In the case of a 404 (not found) status the ui would then get an exception returned from the api which was then incorrectly interpreted as the service being down.
This problem would also be present if for example a non-existent provider was requested.

### Changes proposed in this pull request
A change has been put into the HttpClientWrapper to filter out all 404's and return without raising an exception.
This is now interpreted correctly by the ui and a "Page Not Found" page is shown.
There is now a test which would have flushed out the bug.

### CODE REVIEW

**Comment by MartinBentley** [comment]
[Find showing 503 service unavailable when the api returns 404](https://trello.com/c/cjI2ies3/575-find-showing-503-service-unavailable-when-the-api-returns-404)

File: tests/Unit.Tests/Client/HttpClientWrapperTests.cs [hunk 1]
```diff
@@ -86,6 +86,24 @@ public void GetAsync_SearchAndCompareApiException()
             act.Should().Throw<SearchAndCompareApiException>().WithMessage(msg);
         }
 
+        [Test]
+        public void GetAsync_SearchAndCompareApi404()
+        {
+            var ub = new UriBuilder("test");
+            var uri = ub.Uri;
```
Comment (timabell) [inline tests/Unit.Tests/Client/HttpClientWrapperTests.cs]: minor: I'd inline `ub`

File: tests/Unit.Tests/Client/HttpClientWrapperTests.cs [hunk 2]
```diff
@@ -86,6 +86,24 @@ public void GetAsync_SearchAndCompareApiException()
...
+
+            mockMessageHandler
+                .Protected()
+                .Setup<Task<HttpResponseMessage>>("SendAsync", ItExpr.IsAny<HttpRequestMessage>(), ItExpr.IsAny<CancellationToken>())
+                .ReturnsAsync(new HttpResponseMessage { StatusCode = HttpStatusCode.NotFound });
```
Comment (timabell) [inline tests/Unit.Tests/Client/HttpClientWrapperTests.cs]: question: can we DRY this setup? repeats a lot in the file, maybe extract a method

File: tests/Unit.Tests/Client/HttpClientWrapperTests.cs [hunk 3]
```diff
@@ -86,6 +86,24 @@ public void GetAsync_SearchAndCompareApiException()
...
+
+            Func<Task> act = async () => await sut.GetAsync(uri);
+
+            act.Should().NotThrow<SearchAndCompareApiException>();
```
Comment (timabell) [inline tests/Unit.Tests/Client/HttpClientWrapperTests.cs]: question: this tests that it fails to fail, but can we also test that it's behaving correctly? I can't see from this what the expected behaviour is. Should it throw a different exception, or no exception, or is there something else we can check to validate the response?

Comment (MartinBentley) [reply tests/Unit.Tests/Client/HttpClientWrapperTests.cs]: What it's testing is that is doesn't throw an exception on a 404 which is the expected behaviour.
This test specifically targets the fixed bug and would have flushed it out if it had previously existed. The happy path tests already exist.

Comment (timabell) [reply tests/Unit.Tests/Client/HttpClientWrapperTests.cs]: I don't think it shows the expected behaviour when there's a 404, just the unwanted behaviour we previously had

File: tests/Unit.Tests/Client/HttpClientWrapperTests.cs (hunk 2 above)
Comment (MartinBentley) [reply tests/Unit.Tests/Client/HttpClientWrapperTests.cs]: Done

File: tests/Unit.Tests/Client/HttpClientWrapperTests.cs (hunk 3 above)
Comment (MartinBentley) [reply tests/Unit.Tests/Client/HttpClientWrapperTests.cs]: Ok new test added

**Pull request merged**